bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/handler/FreeDeviceHandler.java

```diff
@@ -0,0 +1,145 @@
+import org.openhab.core.library.CoreItemFactory;
+import org.openhab.core.library.unit.SIUnits;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.builder.ChannelBuilder;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ * @author Gaël L'hopital - Initial contribution
+ */
+@NonNullByDefault
+public abstract class FreeDeviceHandler extends HostHandler {
+    private final Logger logger = LoggerFactory.getLogger(FreeDeviceHandler.class);
+    protected long uptime = -1;
+
+    public FreeDeviceHandler(Thing thing, ZoneId zoneId) {
+        super(thing, zoneId);
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+        try {
+            while (!checkBridgeHandler()) {
+                Thread.sleep(1000);
+            }
+            initializeChannels();
+        } catch (InterruptedException e) {
+            logger.info("Interrupted during initialization", e);
+        }
+    }
+
+    @Override
+    protected void internalPoll() throws FreeboxException {
+        super.internalPoll();
+        fetchSystemConfig();
+    }
+
+    protected abstract DeviceConfig getDeviceConfig() throws FreeboxException;
+
+    protected abstract void internalCallReboot() throws FreeboxException;
+
+    private void initializeChannels() {
+        List<Channel> channels = new ArrayList<>(getThing().getChannels());
+
+        try {
+            DeviceConfig systemConfig = getDeviceConfig();
+            List<Sensor> sensors = systemConfig.getAllSensors();
+            sensors.forEach(sensor -> {
+                ChannelUID sensorId = new ChannelUID(thing.getUID(), PROPERTY_SENSORS, sensor.getId());
+                if (channels.stream().noneMatch(c -> c.getUID().equals(sensorId))) {
+                    ChannelBuilder channelBuilder = ChannelBuilder.create(sensorId).withLabel(sensor.getName());
+                    if (sensor.getKind() == SensorKind.FAN) {
+                        channels.add(channelBuilder.withAcceptedItemType(CoreItemFactory.NUMBER)
+                                .withType(new ChannelTypeUID(BINDING_ID + ":fanspeed")).build());
+                    } else if (sensor.getKind() == SensorKind.TEMP) {
+                        channels.add(channelBuilder.withAcceptedItemType("Number:Temperature")
+                                .withType(new ChannelTypeUID(BINDING_ID + ":temperature")).build());
+                    }
+                }
+            });
+            updateThing(editThing().withChannels(channels).build());
+        } catch (FreeboxException e) {
+            logger.info("Error getting list of optional channels : {}", e.getMessage());
```

Comment:
See above comments. Logging levels and when to use them are covered here:
https://www.openhab.org/docs/developer/guidelines.html#f-logging